src/ja/server/scheduler/default_policies.py

```diff
@@ -29,51 +45,129 @@ class DefaultJobDistributionPolicyBase(JobDistributionPolicy, ABC):
         """
 
+    def _get_job_allocation(self, job: Job) -> ResourceAllocation:
+        return ResourceAllocation(job.docker_constraints.cpu_threads, job.docker_constraints.memory, 0)
+
+    def _check_machine_feasible(self, job: Job, machine: WorkMachine, free_only: bool) -> bool:
+        """
+        Check whether the work machine has enough resources to run the given job.
+        @param free_only If True, only the free resources on the machine are taken into consideration.
+        @return True if feasible, False otherwise.
+        """
+        if free_only:
+            return machine.resources.allocate(self._get_job_allocation(job), test_only=True)
+        return not (machine.resources.total_resources - self._get_job_allocation(job)).is_negative()
+
     def assign_machine(self,
-                       job: Job,
+                       job: DatabaseJobEntry,
                        distribution: ServerDatabase.JobDistribution,
                        available_machines: List[WorkMachine]) -> Optional[Tuple[WorkMachine, List[Job]]]:
-        # Should iterate over all available machines and select the best one depending on _assign_machine_cost
-        pass
+        chosen_machine: Tuple[WorkMachine, List[Job]] = None
+        chosen_cost: float = 0
+
+        for machine in available_machines:
+            job_entries = \
+                list(filter(lambda e: ((e.assigned_machine.uid if e.assigned_machine else None) == machine.uid),
+                     distribution))
+
+            result = self._assign_machine_cost(job, machine, job_entries)
+            if result:
+                (cost, preempted) = result
+                if not chosen_machine or cost < chosen_cost:
+                    chosen_cost = cost
+                    chosen_machine = (machine, preempted)
+        return chosen_machine
 
 
 class DefaultNonPreemptiveDistributionPolicy(DefaultJobDistributionPolicyBase):
+    _cpu_threads_w = 2.0
+    _memory_w = 1.0
     """
     Default distribution policy for non-preemptive scheduling.
     """
+    def __init__(self, cost_function: CostFunction):
+        """
+        Initialize the distribution policy.
+        @param cost_function The cost function to use to determine urgent jobs.
+        """
+        self._cost_func = cost_function
+
     def _assign_machine_cost(self,
-                             job: Job,
+                             job: DatabaseJobEntry,
                              machine: WorkMachine,
-                             existing_jobs: List[Job]) -> Optional[Tuple[int, List[Job]]]:
-        pass
+                             existing_jobs: List[DatabaseJobEntry]) -> Optional[Tuple[float, List[Job]]]:
+        if not self._check_machine_feasible(job.job, machine, free_only=True):
+            return None
+        if machine.resources.free_resources.swap < job.job.docker_constraints.memory:
+            if self._cost_func.calculate_cost(job) > self._cost_func.preempting_threshold:
```

Comment:
No, I am first trying to schedule non preemptively, although it doesn't really matter ...

Reply:
Actually you're right, this doesn't make sense here.

Reply:
Huh, I remember why I did it this way: I am always trying to schedule jobs according to the non-preemptive policy. This makes it easier to implement the other policies because then the other policies do not need to consider the case of non-preemptive scheduling, for ex.: If you have two machines where you can schedule a preempting job, and you don't need to preempt any jobs on either of the two machines, then if we directly schedule with the preemptive policy, the preemptive policy will need to fall back to non-preemptive cost calculation, because otherwise both machines will have the same cost (both have no jobs that we want to pause).

Reply:
I think I understand what you mean.